### PR TITLE
Remove further instances of the ExcInvalidConstructorCall exception.

### PR DESCRIPTION
**Pull request by bangerth**
In the same spirit as #3011 and #3015. Part of #610.

### CODE REVIEW

File: source/lac/sparsity_pattern.cc [hunk 1]
```diff
@@ -226,15 +226,14 @@ SparsityPattern &
...
-  Assert (colnums == 0, ExcInvalidConstructorCall());
-  Assert (rows == 0, ExcInvalidConstructorCall());
-  Assert (cols == 0, ExcInvalidConstructorCall());
+  Assert (s.rows==0 && s.cols==0 && s.rowstart==0 && s.colnums==0,
+          ExcMessage("This operator can only be called if the provided argument "
+                     "is the sparsity pattern for an empty matrix. This operator can "
+                     "not be used to copy a non-empty sparsity pattern."));
+
+  Assert (rows==0 && cols==0 && rowstart==0 && cols==0,
+          ExcMessage("This operator can only be called if the current object is"
```
Comment (drwells) [inline source/lac/sparsity_pattern.cc]: missing a trailing space

File: source/lac/chunk_sparsity_pattern.cc [hunk 2]
```diff
@@ -97,8 +99,14 @@ ChunkSparsityPattern::~ChunkSparsityPattern ()
...
 {
-  Assert (s.rows == 0, ExcInvalidConstructorCall());
-  Assert (s.cols == 0, ExcInvalidConstructorCall());
+  Assert (s.rows==0 && s.cols==0,
+          ExcMessage("This operator can only be called if the provided argument "
+                     "is the sparsity pattern for an empty matrix. This operator can "
+                     "not be used to copy a non-empty sparsity pattern."));
+
+  Assert (rows==0 && cols==0,
+          ExcMessage("This operator can only be called if the current object is"
```
Comment (drwells) [inline source/lac/chunk_sparsity_pattern.cc]: missing a trailing space

**Comment by drwells** [comment]
Pedantic English comment: of course I can call the constructor (no good will come of it); the messages should be "This constructor should only be called if..."

**Comment by bangerth** [comment]
I've fixed the two whitespace issues.

Regarding the grammar thingy: This turns out to be quite widespread, actually:
```
/dealii> grep -r 'can only be called' include/ source/
include/deal.II/lac/sparse_matrix_ez.templates.h:          ExcMessage("This constructor can only be called if the provided argument "
include/deal.II/lac/sparse_matrix_ez.templates.h:          ExcMessage("This operator can only be called if the provided right "
include/deal.II/lac/sparsity_pattern.h:     * This function can only be called for entries for which is_valid_entry()
include/deal.II/lac/sparsity_pattern.h:     * can only be called for entries for which is_valid_entry() is true.
include/deal.II/lac/chunk_sparsity_pattern.h:     * can only be called for entries for which is_valid_entry() is true.
include/deal.II/lac/dynamic_sparsity_pattern.h:   * Number of entries in a specific row. This function can only be called if
include/deal.II/lac/sparse_matrix.templates.h:          ExcMessage("This constructor can only be called if the provided argument "
include/deal.II/lac/sparse_matrix.templates.h:          ExcMessage("This operator can only be called if the provided right "
include/deal.II/lac/chunk_sparse_matrix.templates.h:          ExcMessage("This constructor can only be called if the provided argument "
include/deal.II/lac/chunk_sparse_matrix.templates.h:          ExcMessage("This operator can only be called if the provided right "
include/deal.II/grid/tria_accessor.templates.h:          ExcMessage("is_locally_owned() can only be called on active cells!"));
include/deal.II/grid/tria_accessor.templates.h:          ExcMessage("is_ghost() can only be called on active cells!"));
include/deal.II/grid/tria_accessor.templates.h:          ExcMessage("is_artificial() can only be called on active cells!"));
include/deal.II/grid/tria_accessor.templates.h:          ExcMessage("subdomain_id() can only be called on active cells!"));
include/deal.II/base/index_set.h:   * This function can only be called if the index set does not yet contain
include/deal.II/base/index_set.h:          ExcMessage ("This function can only be called if the current "
include/deal.II/dofs/dof_accessor.templates.h:          ExcMessage ("This function can only be called for active fe indices"));
include/deal.II/dofs/dof_accessor.h:   * This iterator can only be called for one-dimensional triangulations.
include/deal.II/dofs/dof_handler.h:  Assert(has_level_dofs(), ExcMessage("n_dofs(level) can only be called after distribute_mg_dofs()"));
source/lac/sparsity_pattern.cc:          ExcMessage("This constructor can only be called if the provided argument "
source/lac/sparsity_pattern.cc:          ExcMessage("This operator can only be called if the provided argument "
source/lac/sparsity_pattern.cc:          ExcMessage("This operator can only be called if the current object is "
source/lac/chunk_sparsity_pattern.cc:          ExcMessage("This constructor can only be called if the provided argument "
source/lac/chunk_sparsity_pattern.cc:          ExcMessage("This operator can only be called if the provided argument "
source/lac/chunk_sparsity_pattern.cc:          ExcMessage("This operator can only be called if the current object is "
source/lac/block_sparsity_pattern.cc:          ExcMessage("This constructor can only be called if the provided argument "
source/lac/dynamic_sparsity_pattern.cc:          ExcMessage("This constructor can only be called if the provided argument "
source/lac/dynamic_sparsity_pattern.cc:          ExcMessage("This operator can only be called if the provided argument "
source/lac/dynamic_sparsity_pattern.cc:          ExcMessage("This operator can only be called if the current object is"
source/grid/tria_accessor.cc:          ExcMessage("set_subdomain_id() can only be called on active cells!"));
source/base/quadrature_lib.cc:        Assert (false, ExcMessage ("This constructor can only be called with either "
```
I think saying "should only be called" is also not quite correct. That would imply a *suggestion*, not a *command*. What would you think of the wording "must only be called"?

**Comment by bangerth** [comment]
(If you allow me, I would also like to address the grammar question in a separate pull request to keep separate concerns separate.)

**Comment by drwells** [comment]
Yes, this is not a serious concern.

**Comment by drwells** [comment]
Writing "must only be called" would be more accurate, though I prefer "should only be called" since it sounds more polite to me.

English is hard. Maybe we should take an opinion poll.

**Comment by bangerth** [comment]
Nah, you and @jppelteret are the only native English speakers among us. You two come up with a solution and I will implement it.

**Comment by jppelteret** [comment]
One could write "It is only possible to call...". That is both polite and factual.

P.S. Oftens I cannot be Englishing very good.

**Comment by jppelteret** [comment]
Or "Calling <XYZ> is only valid if...".

**Comment by Rombur** [comment]
Image

**Comment by jppelteret** [comment]
From the [forum guidelines](https://groups.google.com/forum/#!topic/dealii/GRZMUTLIm2I):

> No contribution is too small...

I like to think that this has universal application :-)

**Comment by bangerth** [comment]
We're no doubt doing god's work in figuring out how best to phrase this!

**Comment by bangerth** [comment]
Since "It is only possible to call this [...]" falls into the same trap @drwells originally mentioned, I'm tempted to go with your second option, "Calling [...] is only valid...".

Are we agreed that this is the way to go?

**Comment by drwells** [comment]
okay, more bikeshedding. I am probably outvoted and that is fine.

I don't like that use of the word valid here: it just sounds wrong to me (I do not have a rigorous explanation for why). I still think that 'should' is the right way here.

**Comment by jppelteret** [comment]
@drwells Possible synonyms of valid (assuming that contextually it is the correct message one wants to convey): legal, legitimate, correct, allowed. And with that, I've run out of ideas. Do you find any of these more palatable?

**Comment by drwells** [comment]
How about, to split this into two cases (and further procrastinate on writing lecture notes):

```cpp
ExcMessage("is_ghost() has a meaningful result only if it is called on an active cell."));
```
rather than
```cpp
ExcMessage("is_ghost() can only be called on active cells!"));
```
and
```cpp
Assert (s.rows==0 && s.cols==0,
        ExcMessage("This constructor makes sense only if the provided argument "
                   "is the sparsity pattern for an empty matrix. This constructor can "
                   "not be used to copy-construct a non-empty sparsity pattern."));
```
instead of
```cpp
Assert (s.rows==0 && s.cols==0,
        ExcMessage("This constructor can only be called if the provided argument "
                   "is the sparsity pattern for an empty matrix. This constructor can "
                   "not be used to copy-construct a non-empty sparsity pattern."));

```

so I would prefer "meaningful result" for functions and "makes sense" for constructors instead of just 'is valid'.

**Comment by bangerth** [comment]
I think "has a meaningful result" and "makes sense only" really doesn't convey the fact that you won't pass the exception. You don't get *any* kind of result, not even a meaningless one.

**Comment by drwells** [comment]
Hmm, good points. It looks like 'is valid' is the best choice.

**Comment by davydden** [comment]
So, it looks like weeee have a winer -- `is valid` 🎉

**Pull request merged**